[ao] Added constant channel check to Outlier Detector

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81369
* #81382
* #81252
* #81251
* #81250
* __->__ #81249
* #81174

Summary: The current Outlier detector does a good job of finding whether
data distributions passing through layers have outliers. However,
suppose we have a completely constant channel. The outlier detector
would not detect it as an outlier, but that is still something we want
to highlight because a constant channel usually is a result of a bad
configuration or something really wrong with the data.

To address this there are two additions to the outlier detector that
this commit makes:
- The first is to add whether there are any constant batches at all and
let the user know in the text report
- The second is to let the user know the number of total constant
batches found for each channel, so they can figure out if there are any
unnecessary channels present.

The exisiting outlier detector tests were modified to do a quick check
for this feature.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Added constant channel check to Outlier Detector

Summary: The current Outlier detector does a good job of finding whether
data distributions passing through layers have outliers. However,
suppose we have a completely constant channel. The outlier detector
would not detect it as an outlier, but that is still something we want
to highlight because a constant channel usually is a result of a bad
configuration or something really wrong with the data.

To address this there are two additions to the outlier detector that
this commit makes:
- The first is to add whether there are any constant batches at all and
let the user know in the text report
- The second is to let the user know the number of total constant
batches found for each channel, so they can figure out if there are any
unnecessary channels present.

The exisiting outlier detector tests were modified to do a quick check
for this feature.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added constant channel check to Outlier Detector"


Summary: The current Outlier detector does a good job of finding whether
data distributions passing through layers have outliers. However,
suppose we have a completely constant channel. The outlier detector
would not detect it as an outlier, but that is still something we want
to highlight because a constant channel usually is a result of a bad
configuration or something really wrong with the data.

To address this there are two additions to the outlier detector that
this commit makes:
- The first is to add whether there are any constant batches at all and
let the user know in the text report
- The second is to let the user know the number of total constant
batches found for each channel, so they can figure out if there are any
unnecessary channels present.

The exisiting outlier detector tests were modified to do a quick check
for this feature.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added constant channel check to Outlier Detector"


Summary: The current Outlier detector does a good job of finding whether
data distributions passing through layers have outliers. However,
suppose we have a completely constant channel. The outlier detector
would not detect it as an outlier, but that is still something we want
to highlight because a constant channel usually is a result of a bad
configuration or something really wrong with the data.

To address this there are two additions to the outlier detector that
this commit makes:
- The first is to add whether there are any constant batches at all and
let the user know in the text report
- The second is to let the user know the number of total constant
batches found for each channel, so they can figure out if there are any
unnecessary channels present.

The exisiting outlier detector tests were modified to do a quick check
for this feature.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added constant channel check to Outlier Detector"


Summary: The current Outlier detector does a good job of finding whether
data distributions passing through layers have outliers. However,
suppose we have a completely constant channel. The outlier detector
would not detect it as an outlier, but that is still something we want
to highlight because a constant channel usually is a result of a bad
configuration or something really wrong with the data.

To address this there are two additions to the outlier detector that
this commit makes:
- The first is to add whether there are any constant batches at all and
let the user know in the text report
- The second is to let the user know the number of total constant
batches found for each channel, so they can figure out if there are any
unnecessary channels present.

The exisiting outlier detector tests were modified to do a quick check
for this feature.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1079,11 +1085,13 @@ def _calculate_outlier_info(
             "outliers_detected" : list of bools per channel that are true if it is considered an outlier
             "sufficient_batches": if o_r was >= statistical_threshold:
                 where o_r = counted_batches / total_batches
+            "constant_channels": if non_constant_per_channel != total_batches for each channel
```
What is `non_constant_per_channel`? I couldn't find any references to this anywhere

andrewor14 replied:
Also is this a list of bools or a list of counts? Should clarify in comments

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1164,6 +1181,8 @@ def _generate_info_dict(self, model: GraphModule) -> Dict[str, Dict]:
                     self.OUTLIER_KEY: outlier_calcs[self.OUTLIER_KEY],
                     self.SUFFICIENT_BATCHES_KEY: outlier_calcs[self.SUFFICIENT_BATCHES_KEY],
+                    self.CONSTANT_CHANNEL_KEY: outlier_calcs[self.CONSTANT_CHANNEL_KEY],
+                    self.CONSTANT_COUNTS_KEY: channel_batch_cnts,
```
Do we need both of these? It seems if `num_constant_batches` == 0 then `has_constant_batches` will be False? Correct me if I'm misunderstanding something but the latter seems redundant.

vspenubarthi replied:
you are right that if `num_constant_batches == 0` then has_constant_batches is False, i just had done this to make the logic of generating the text report easier.

However, this is redundant data, so I will make this fix by removing the boolean value.

vspenubarthi committed:
Update on "[ao] Added constant channel check to Outlier Detector"


Summary: The current Outlier detector does a good job of finding whether
data distributions passing through layers have outliers. However,
suppose we have a completely constant channel. The outlier detector
would not detect it as an outlier, but that is still something we want
to highlight because a constant channel usually is a result of a bad
configuration or something really wrong with the data.

To address this there are two additions to the outlier detector that
this commit makes:
- The first is to add whether there are any constant batches at all and
let the user know in the text report
- The second is to let the user know the number of total constant
batches found for each channel, so they can figure out if there are any
unnecessary channels present.

The exisiting outlier detector tests were modified to do a quick check
for this feature.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added constant channel check to Outlier Detector"


Summary: The current Outlier detector does a good job of finding whether
data distributions passing through layers have outliers. However,
suppose we have a completely constant channel. The outlier detector
would not detect it as an outlier, but that is still something we want
to highlight because a constant channel usually is a result of a bad
configuration or something really wrong with the data.

To address this there are two additions to the outlier detector that
this commit makes:
- The first is to add whether there are any constant batches at all and
let the user know in the text report
- The second is to let the user know the number of total constant
batches found for each channel, so they can figure out if there are any
unnecessary channels present.

The exisiting outlier detector tests were modified to do a quick check
for this feature.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
